autorally_core/src/autorally_chassis/AutoRallyChassis.cpp

```diff
@@ -0,0 +1,583 @@
+  
+  double commandRate = 0.0;
+  double chassisCommandMaxAge = 0.0;
+  double runstopMaxAge = 0.0;
+
+  if(!nhPvt.getParam("commandRate", commandRate) ||
+     !nhPvt.getParam("commandMaxAge", chassisCommandMaxAge) ||
+     !nhPvt.getParam("runstopMaxAge", runstopMaxAge))
+  {
+    NODELET_ERROR_STREAM(getName() << " could not get all startup params");
+  }
+  chassisCommandMaxAge_ = ros::Duration(chassisCommandMaxAge);
+  runstopMaxAge_ = ros::Duration(runstopMaxAge);
+
+  chassisStatePub_ = nh.advertise<autorally_msgs::chassisState>
+                     ("chassisState", 1);
+  wheelSpeedsPub_ = nh.advertise<autorally_msgs::wheelSpeeds>
+                     ("wheelSpeeds", 1);
+  chassisCommandPub_ = nh.advertise<autorally_msgs::chassisCommand>
+                     ("RC/chassisCommand", 1);
+
+  for (auto& mapIt : chassisCommands_)
+  {
+    std::string topic = mapIt.first+"/chassisCommand";
+    ros::Subscriber sub = nh.subscribe(topic, 1,
+                                       &AutoRallyChassis::chassisCommandCallback, this);
+    chassisCommandSub_[mapIt.first] = sub;
+  }
+  runstopSub_ = nh.subscribe("/runstop", 5, &AutoRallyChassis::runstopCallback, this);  
+  
+  //callback for serial data from chassis
+  serialPort_.registerDataCallback(boost::bind(&AutoRallyChassis::chassisFeedbackCallback, this));
+
+  chassisControlTimer_ = nh.createTimer(ros::Rate(commandRate),
+                      &AutoRallyChassis::setChassisActuators, this);
+}
+
+//subscribe to a one topic for every chassis commander listed in the chassis commander priorities file
+void AutoRallyChassis::chassisCommandCallback(
+                     const autorally_msgs::chassisCommandConstPtr& msg)
+{
+  std::map<std::string, autorally_msgs::chassisCommand>::iterator mapIt;
+  if((mapIt = chassisCommands_.find(msg->sender)) == chassisCommands_.end())
+  {
+    NODELET_ERROR_STREAM("AutoRallyChassis: Unknown controller " <<
+                         msg->sender <<
+                         " attempting to control chassis, please add entry " <<
+                         " to chassisCommandPriorities.yaml");
+  } else
+  {
+    mapIt->second = *msg;
+  }
+}
+
+void AutoRallyChassis::chassisFeedbackCallback()
+{
+  //Any variables accessed in here and in other places in the code need to be mutex'd as this fires in a different
+  //thread than the main thread. This is also a pretyt long callback, and ROS can shutdown underneath us, so check
+  //if ROS system is still running any time anything ROS is used
+
+  serialPort_.lock();
+  size_t startPosition = serialPort_.m_data.find_first_of('#');
+  size_t endPosition = serialPort_.m_data.find_first_of('\n', startPosition);
+  serialPort_.unlock();
+  std::string data = "";
+
+  //retrieve and process all available messages from the chassis
+  while(endPosition != std::string::npos)
+  {
+    //make sure we have start and end position
+    serialPort_.lock();
```

Comment:
If there is the potential to erase data from this buffer in another part of the code, the mutex needs to stay locked from above.